File: src/go11/day11.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"time"
)

type matrix [][]Octopus
type Octopus struct {
	energy  uint8
	flashed bool
}

func BuildMatrix(lines []string) matrix {
	m := make([][]Octopus, len(lines))
	for i, l := range lines {
		l = strings.TrimSpace(l)
		m[i] = make([]Octopus, len(l))
		for j, c := range l {
			m[i][j] = Octopus{energy: uint8(c - '0'), flashed: false}
		}
	}
	return m
}

func increase_energy(m matrix) {
	for i := range m {
		for j := range m[i] {
			m[i][j].energy += 1
		}
	}
}

func increase_neighbours_energy(m matrix, x, y int) {
	for j := y - 1; j <= y+1; j++ {
		for i := x - 1; i <= x+1; i++ {
			if j >= 0 && j < len(m) && i >= 0 && i < len(m[j]) {
				m[j][i].energy += 1
			}
		}
	}
	m[y][x].energy -= 1
}

func flash(m matrix) int {
	continue_flashing := true
	flashed := 0
	for continue_flashing {
		continue_flashing = false
		for y := range m {
			for x := range m[y] {
				if m[y][x].energy > 9 && !m[y][x].flashed {
					m[y][x].flashed = true
					flashed++
					continue_flashing = true
					increase_neighbours_energy(m, x, y)
				}
			}
		}
	}
	return flashed
}

func clear_flashed(m matrix) bool {
	all_flashed := true
	for y := range m {
		for x := range m[y] {
			if m[y][x].flashed {
				m[y][x].flashed = false
				m[y][x].energy = 0
			} else {
				all_flashed = false
			}
		}
	}
	return all_flashed
}

func Part1(input string) int {
	lines := strings.Split(strings.TrimSuffix(input, "\n"), "\n")
	m := BuildMatrix(lines)

	res := 0
	for i := 1; i <= 100; i++ {
		increase_energy(m)
		res += flash(m)
		clear_flashed(m)
	}
	return res
}

func Part2(input string) int {
	lines := strings.Split(strings.TrimSuffix(input, "\n"), "\n")
	m := BuildMatrix(lines)

	for i := 1; ; i++ {
		increase_energy(m)
		flash(m)
		if clear_flashed(m) {
			return i
		}
	}
}

func main() {
	content, _ := ioutil.ReadFile("../../inputs/day11.txt")

	start := time.Now()
	fmt.Println("part1: ", Part1(string(content)))
	fmt.Println(time.Since(start))

	start = time.Now()
	fmt.Println("part2: ", Part2(string(content)))
	fmt.Println(time.Since(start))
}

File: src/day11.rs
#[derive(Debug)]
struct Octopus {
    energy: u32,
    flashed: bool,
}

fn increase_energy(board: &mut Vec<Vec<Octopus>>) {
    for y in 0..board.len() {
        for x in 0..board[y].len() {
            board[y][x].energy += 1;
        }
    }
}

fn increase_neighbors_energy(board: &mut Vec<Vec<Octopus>>, x: usize, y: usize) {
    for j in (y as i32 - 1)..(y + 1 + 1) as i32 {
        for i in (x as i32 - 1)..(x + 1 + 1) as i32 {
            if j >= 0 && j < board.len() as i32 && i >= 0 && i < board[j as usize].len() as i32 {
                board[j as usize][i as usize].energy += 1;
            }
        }
    }
    board[y][x].energy -= 1;
}

fn clear_flash(board: &mut Vec<Vec<Octopus>>) -> bool {
    let mut all_falshed = true;
    for y in 0..board.len() {
        for x in 0..board[y].len() {
            if board[y][x].flashed {
                board[y][x].energy = 0;
                board[y][x].flashed = false;
            } else {
                all_falshed = false;
            }
        }
    }
    return all_falshed;
}

fn flash(board: &mut Vec<Vec<Octopus>>) -> u32 {
    let mut continue_flashing = true;
    let mut flashed = 0;
    while continue_flashing {
        continue_flashing = false;
        for y in 0..board.len() {
            for x in 0..board[y].len() {
                if board[y][x].energy > 9 && !board[y][x].flashed {
                    board[y][x].flashed = true;
                    flashed += 1;
                    continue_flashing = true;
                    increase_neighbors_energy(board, x, y);
                }
            }
        }
    }
    return flashed;
}

pub fn part1(input: String) -> i64 {
    let mut board: Vec<Vec<Octopus>> = input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| Octopus {
                    energy: c.to_digit(10).unwrap(),
                    flashed: false,
                })
                .collect()
        })
        .collect();

    let mut res = 0;
    for _ in 1..101 {
        increase_energy(&mut board);
        let flashed = flash(&mut board);
        res += flashed;
        clear_flash(&mut board);
    }

    return res as i64;
}

pub fn part2(input: String) -> i64 {
    let mut board: Vec<Vec<Octopus>> = input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| Octopus {
                    energy: c.to_digit(10).unwrap(),
                    flashed: false,
                })
                .collect()
        })
        .collect();

    let mut i = 1;
    loop {
        increase_energy(&mut board);
        flash(&mut board);
        let all_flashed = clear_flash(&mut board);
        if all_flashed {
            return i;
        }
        i += 1
    }
}
